### PR TITLE
Add in-app update prompts

### LINKED ISSUE #49: Add Google Play app update available notifications
Add notifications to nag and/or force users to update to the latest version of the app.  The priority of each release can be set during our CI process.

My proposed update forcing strategy is below. The intent is to limit the number of older releases we have to maintain compatibility with, so this is intentionally aggressive.

 - If the app update priority is 5, it has to be installed immediately
 - If the app update priority is 4, and the update staleness (days since update was released) is <=1 day, show an optional update prompt.  If the update staleness is at least 1 day, immediately install the update.
 - If the update priority is 3 and the update staleness is at least 4 days, show an optional update prompt.
 - If the update staleness is at least 14 days regardless of priority, show an optional prompt.
 - If the update staleness is at least 21 days regardless of priority, immediately install the update. 

App update priorities, inspired by [Google's suggestions](https://developer.android.com/guide/playcore/in-app-updates/kotlin-java#update-priority):

> - Minor UI improvements: Low-priority update; request neither an optional update nor an immediate update. Update only when the user isn't interacting with your app.
> - Performance improvements: Medium-priority update; request an optional update.
> - Critical security update: High-priority update; request an immediate update.

| Update priority | Description | Examples | Update timeline for users |
| --- | --- |--- | --- |
| 0/1/2 | Very low  or low priority | UI touchups that don't impact functionality, releases with options to opt-in to beta features | No prompt initially. Optional prompt starting 14 days after release. Immediate update 21 days after release. |
| 3 | Medium priority |  Medium-priority bug fixes, performance improvements, non-time-sensitive feature launches | No prompts for the first 3 days. Optional starting 4 days after release. Immediate update 21 days after release. |
| 4 | High priority or time-sensitive | High priority bug fixes, time-sensitive feature launches | Optional for the first 24 hours, then immediate. |
| 5 | Critical bug fixes | Crashes/bugs impacting major features, urgent vulnerabilities | Immediate update required. |

- Be sure to update our data collection info on the Play Console after merging this feature: https://developer.android.com/guide/playcore/in-app-updates#in-app_updates

### PR DESCRIPTION
**Pull request by evan10s**
This update adds in-app update prompts via Google Play. See #49 for more information about when update prompts will appear.

The Google Play immediate update method is always used, which means users get kicked out of the app for less than a minute so that the new update can be downloaded and installed.

Users may see an optional or required update prompt.

The optional updated prompt is a bottom sheet, which will appear on every app launch but can be dismissed:

The required update prompt overtakes the entire app and cannot be dismissed:

To clean up MainActivity.kt some, I refactored a few things such as NfcRequired into their own "gate" components.

Closes #49 

TODO:
- [x] UI for less common app update states: update downloaded and update in progress
- [x] Make sure Detekt passes locally
- [x] Write instructions to test update flow
- [x] Devise way to specify update status in CI
- [x] Debug weird flashing issue when app opens
- [x] [Code snippet](https://developer.android.com/guide/playcore/in-app-updates/kotlin-java#immediate) for IAU that's supposed to go in MainActivity `onResume`
- [x] From #49: Be sure to update our data collection info on the Play Console after merging this feature: https://developer.android.com/guide/playcore/in-app-updates#in-app_updates

### CODE REVIEW

File: base/src/main/java/org/robojackets/apiary/base/ui/update/UpdateAvailable.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.unit.dp
+import com.google.android.play.core.ktx.AppUpdateResult
+import org.robojackets.apiary.base.ui.icons.UpdateIcon
+import org.robojackets.apiary.base.ui.util.getActivity
+import se.warting.inappupdate.compose.rememberInAppUpdateState
+import timber.log.Timber
+
+// Just a random number so we can identify our update request later if necessary
+const val UPDATE_REQUEST_CODE = 1999
```
Comment (kberzinch) [inline base/src/main/java/org/robojackets/apiary/base/ui/update/UpdateAvailable.kt:19]: What exactly is this for?

**Review by kberzinch — DISMISSED** [review]

File: app/src/main/java/org/robojackets/apiary/MainActivity.kt [hunk 2]
```diff
@@ -66,7 +57,7 @@ sealed class Screen(
     val navigationDestination: String,
     @StringRes val resourceId: Int,
     val icon: ImageVector,
-    val imgContentDescriptor: String
+    val imgContentDescriptor: String,
```
Comment (JoshOldenburg) [inline app/src/main/java/org/robojackets/apiary/MainActivity.kt:60]: What does the comma do?

**Review by JoshOldenburg — APPROVED** [review]

File: app/src/main/java/org/robojackets/apiary/MainActivity.kt (hunk 2 above)
Comment (evan10s) [reply app/src/main/java/org/robojackets/apiary/MainActivity.kt:60]: Looks like Detekt added this. But it's just putting a trailing comma on every line of the list stylistically.

File: base/src/main/java/org/robojackets/apiary/base/ui/update/UpdateAvailable.kt (hunk 1 above)
Comment (evan10s) [reply base/src/main/java/org/robojackets/apiary/base/ui/update/UpdateAvailable.kt:19]: When you call the start update function, I believe you have to provide it as an identifier so you can get the status of your update request later. I think(?) this is used somewhere just a few lines below

Comment (kberzinch) [reply base/src/main/java/org/robojackets/apiary/base/ui/update/UpdateAvailable.kt:19]: Weird API but ok

**Comment by evan10s** [comment]
Testing instructions:

**Optional update flow**
1. Download version code 18 of the app from this [internal app sharing build link](https://play.google.com/apps/test/RQmA0F99ukw/ahABH4UwYDNSGsZHBxoXVAtggoeRMV_VKHSbStG_3Mrp7eI70e9zVobxxaBG6MyQ2ekvfC2rMdRQ9QJwxKPAWARJOm)
  a. You will need to uninstall the production version first. If you get an error when installing the build, try restarting your phone.
2. Open the app. Sign in, optionally.
3. Close the app.
4. Open [this internal app sharing link](https://play.google.com/apps/test/RQmA0F99ukw/ahABH4UwYELP-erG7uH_Em5SHrb3zEd5Wmb1Sr51tX7tOPQLKv3G46vOKQEcDr-EeT8Br7B-9XMMgYYfYJCvJi3NfO) for version code 19. Do **not** press the Update button!
5. Go to your home screen and re-open the app. You should be able to go through the optional update flow.
  a. Optionally repeat for signed in/out and light/dark mode, if you feel so inclined

**Required update flow**
1. Uninstall any existing versions of the app on your phone.
2. Repeat the above steps, but use these internal app sharing links for the old and new app versions.
  a. Old version - [v18 req](https://play.google.com/apps/test/RQmA0F99ukw/ahABH4UwbcJyFVnRTsTmXT2fSl6-ag5OIBIkmtAz3zaoyJmeNnG6ZezC_NK2_ms-CE4UCbrG7_3UBw40zydQXNIzOA)
  b. New version - [v19 req](https://play.google.com/apps/test/RQmA0F99ukw/ahABH4UwasyVgPKkDWhlek9V2S9bCBPIzIqazShH5FCv-lZxEm0UDRvoPnxg8lxMZm6GCE4IDT6vlYQHWFE9AX0FF2)

**Comment by evan10s** [comment]
Some resources for final verifications:
- Download [latest test build](https://play.google.com/apps/test/RQmA0F99ukw/ahAIvA2zVrJcuWTiqjbqyEFlo0uXZGy3yvPPfkmBZgPuseMCPFwaX-b6N0dJpgrWzN13YVObVLPjYPz1ss6fWfR9sN) for this branch to confirm nothing is broken
- Test builds that will only trigger the optional update flow:
  - 30: [1.1.0-optional-test-1](https://play.google.com/apps/test/RQmA0F99ukw/ahAIvA2zXxWI60Mw6ctxRCaLfZPt4XtW9ldK9-Au-j8ZUbaUvKiN8AcYffCXEkpDotkGjVIux_ADeABtAudsOD96Pk)
  - 31: [1.1.0-optional-test-2](https://play.google.com/apps/test/RQmA0F99ukw/ahAIvA2zWR3UGE8k5pBA76bx-FRIoiMgrpotnv3UAANuuWoKm-EbEBVPOKx9GJyL9VKgHUJl_8nH9vhXKNT0EV9_ca)
- Test builds that will only trigger the required update flow:
  - 40: [1.1.0-required-test-1](https://play.google.com/apps/test/RQmA0F99ukw/ahAIvA2zXJTtNbptmYiwdaafUKqR21h0_vdFH_OGvMTKqM2ITUMucZwylYhCU1yEfcq2fWktVWycPLkOzlEJRs5W3Z)
  - 41: [1.1.0-required-test-2](https://play.google.com/apps/test/RQmA0F99ukw/ahAIvA2zUUCPrzYVN67NdJYwWufQuU1ng7nX7IRDWyhr6_-a8lKkM3ygXQwPQ1X0Czo3ZbTQ9YLOakOH1AKLlguTvH)

**Review by kberzinch — DISMISSED** [review]
30/31/40/41 test builds don't trigger update prompts for me, other than that seems fine

**Review by kberzinch — APPROVED** [review]

**Pull request merged**